lib/auth/userloginstate/generator.go

```diff
@@ -199,38 +208,97 @@ func (g *Generator) addAccessListsToState(ctx context.Context, user types.User,
 		// Grants are inherited if the user is a member of the access list, explicitly or via inheritance.
-		membershipKind, err := accesslists.IsAccessListMember(ctx, user, accessList, g.accessLists, g.accessLists, g.clock)
+		inheritedRoles, inheritedTraits, err := g.handleAccessListMembership(ctx, user, accessList, state)
 		if err != nil {
-			g.log.WithError(err).Warn("checking access list membership")
+			return nil, nil, trace.Wrap(err)
 		}
-		if err == nil && membershipKind != accesslists.MembershipOrOwnershipTypeNone {
-			g.grantRolesAndTraits(accessList.Spec.Grants, state)
-			if membershipKind == accesslists.MembershipOrOwnershipTypeInherited {
-				allInheritedRoles = append(allInheritedRoles, accessList.Spec.Grants.Roles...)
-				for k, values := range accessList.Spec.Grants.Traits {
-					allInheritedTraits[k] = append(allInheritedTraits[k], values...)
-				}
-			}
+		allInheritedRoles = append(allInheritedRoles, inheritedRoles...)
+		for k, values := range inheritedTraits {
+			allInheritedTraits[k] = append(allInheritedTraits[k], values...)
 		}
+
 		// OwnerGrants are inherited if the user is an owner of the access list, explicitly or via inheritance.
-		ownershipType, err := accesslists.IsAccessListOwner(ctx, user, accessList, g.accessLists, g.accessLists, g.clock)
+		inheritedRoles, inheritedTraits, err = g.handleAccessListOwnership(ctx, user, accessList, state)
 		if err != nil {
-			g.log.WithError(err).Warn("checking access list ownership")
+			return nil, nil, trace.Wrap(err)
 		}
-		if err == nil && ownershipType != accesslists.MembershipOrOwnershipTypeNone {
-			g.grantRolesAndTraits(accessList.Spec.OwnerGrants, state)
-			if ownershipType == accesslists.MembershipOrOwnershipTypeInherited {
-				allInheritedRoles = append(allInheritedRoles, accessList.Spec.OwnerGrants.Roles...)
-				for k, values := range accessList.Spec.OwnerGrants.Traits {
-					allInheritedTraits[k] = append(allInheritedTraits[k], values...)
-				}
-			}
+		allInheritedRoles = append(allInheritedRoles, inheritedRoles...)
+		for k, values := range inheritedTraits {
+			allInheritedTraits[k] = append(allInheritedTraits[k], values...)
 		}
 	}
 
 	return allInheritedRoles, allInheritedTraits, nil
 }
 
+// handleAccessListMembership validates the access list and applies the grants and traits from the access list to the user if they are a member of the access list.
+func (g *Generator) handleAccessListMembership(ctx context.Context, user types.User, accessList *accesslist.AccessList, state *userloginstate.UserLoginState) ([]string, map[string][]string, error) {
+	var inheritedRoles []string
+	inheritedTraits := make(map[string][]string)
+
+	membershipKind, err := accesslists.IsAccessListMember(ctx, user, accessList, g.accessLists, g.accessLists, g.clock)
+	if err != nil {
+		g.log.WithError(err).Warn("checking access list membership")
+	}
+	if err == nil && membershipKind != accesslists.MembershipOrOwnershipTypeNone {
```

Comment:
Can we return in the `err != nil` case so all of this code doesn't have to be inside a `err == nil` check?

Reply:
This is based on the previous implementation, we don't actually want to return if there's an error, just warn and continue.

https://github.com/gravitational/teleport/blob/ed46e873212e6e07225ccf25cecfa604851016d3/lib/auth/userloginstate/generator.go#L206

Reply:
If there is an error, this condition will be false, so the entire if statement will be skipped and you'll go straight to line 265: `return inheritedRoles, inheritedTraits, nil`.

I would invert the check so that you can `return inheritedRoles, inheritedTraits, nil` and then the main body of this function doesn't need to be indented in an if.